saviSimulation/savi/jason_processing/behaviour/UxVBehavior.java

```diff
@@ -0,0 +1,240 @@
+    protected PVector speedVector;
+    protected double compasAngle;
+    protected ArrayList<CameraPerception> visibleItems;
+    protected double time;
+    protected Map<String, AgentAction> actionMap;
+
+    //***********************************************************//
+    //I THINK IS BETTER TO HAVE THE ROBOTS ITS DATA AND THE SYNCAGENTSTATE ITS OWN.
+    //IF WE WANT TO IMPLEMENTE MALFUNCTION OF SENSORS, THE INFO RECEIVED IN
+    //SYNCAGENTSTATE IS NOT THE REAL ONE
+    //***********************************************************//
+
+    //-----------------------------------------
+    // METHODS (functions that act on the data)
+    //-----------------------------------------
+
+    /**
+     * Constructor
+     *
+     * @param id
+     * @param type
+     * @param initialPosition
+     */
+    public UxVBehavior(String id, String type, PVector initialPosition, double reasoningCyclePeriod) {
+        // Initialize data values
+        super(reasoningCyclePeriod);
+        this.ID = id;
+        this.type = type;
+        this.initialPosition = initialPosition;
+        this.speedVector = new PVector();
+        this.time = 0;
+        this.compasAngle = 0;
+        this.agentState = new SyncAgentState();
+        this.visibleItems = new ArrayList<CameraPerception>();
+        this.actionMap = new HashMap<>();
+        this.createActionMap();
+        updatePercepts(initialPosition);
+    }
+
+    public ArrayList<CameraPerception> getVisibleItems() {
+        return visibleItems;
+    }
+
+    public double getCompassAngle() {
+        return this.compasAngle;
+    }
+
+    /**
+     * update
+     * process actions from the queue, update the UAS state variable and set the new perceptions
+     */
+    public void update(UxV uxv, double simTime, int perceptionDistance, List<WorldObject> worldObjects) {
+        List<String> actionsToExecute = agentState.getAllActions();
+
+        //Process actions to update speedVector & compassAngle
+        processAgentActions(actionsToExecute);
+
+        PVector newPos = this.updatePosition(uxv.getPosition(), simTime);
+        uxv.setPosition(newPos);
+
+        //Calculate visible items
+        this.visibleItems.clear();
+
+        //Calculate objects detected with camera
+        visibleItems.addAll(objectDetection(uxv.getPosition(), worldObjects, perceptionDistance));
+
+        //Update percepts
+        updatePercepts(uxv.getPosition());
+    }
+
+    protected PVector updatePosition(PVector currentPosition, double simTime) {
+        //Update simTime
+        double timeElapsed = simTime - this.time; //elapsed time since last update
+        this.time = simTime;
+
+        //Calculate new x,y position
+        double cosv = Math.cos(this.compasAngle);
+        double sinv = Math.sin(this.compasAngle);
+        PVector deltaPos = new PVector((float) (cosv * this.speedVector.x * timeElapsed), (float) (sinv * this.speedVector.y * timeElapsed), this.speedVector.z);
```

Comment:
The reason why the speedVector change is not causing a problem is that there's an issue here that's "cancelling out" the change.
Using it properly would make this line:
deltaPos = new PVector((float) (this.speedVector.x * timeElapsed), (float) (this.speedVector.y * timeElapsed), (float) (this.speedVector.z * timeElapsed));

I suggest we revert to a speedValue scalar. 
Right now we're maintaining a speedvalue which is the projection of speedVector on the x/y plane + compassAngle (which we need anyway in several places) + vertical speed. We discussed this and decided that we didn't want another angle (pitch). So we just put in vertical speed, and as a result our representation of speed is a bit bastardized, a hybrid of polar and cartesian....

Reply:
Understood! Reverting the change!

Reply:
Note that if we move to representing speed as a single vector we still can't get rid of angles because otherwise when we stop speedVector =(0,0,0) we lose the information of which direction we're facing.

Reply:
Should I be using the timeElapsed when calculating the z component? Before my changes, the z component is calculated by:
```
PVector newpos = new PVector((float) (cosv*this.speedVal*timeElapsed), (float) (sinv*this.speedVal*timeElapsed), 0);	

//Calculate new altitude	
newpos.z += verticalSpeedVal;	

if(newpos.z < 0) {	
    newpos.z=0;	
}	

uav.setPosition(uav.getPosition().add(newpos));
```

but should it be:

`newpos.z += verticalSpeedVal * timeElapsed`

Reply:
yeah I guess vertical movement should also vary w.r.t. time elapsed